iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp

```diff
@@ -45,193 +44,435 @@ namespace iree_compiler {
-static constexpr int kIndexPushConstant = 1;
-static constexpr int kIndexWorkGroupId = 2;
-static constexpr int kIndexWorkGroupCount = 3;
-static constexpr int kIndexWorkGroupSize = 4;
+// NOTE: HALDispatchABI and the associated conversion patterns should live under
+// iree/compiler/Dialect/HAL/Target/LLVM/ instead of here as they have nothing
+// to do with linalg. If we need to use the patterns in this conversion we can
+// expose a populate*Patterns() function to access them without needing them
+// defined here.
+
+// Utility for accessing the IREE HAL dispatch function ABI values.
+// All accesses should route through this vs directly manipulating LLVM function
+// arguments so that we can adjust the ABI over time and support multiple
+// versions in the same compiled output.
+class HALDispatchABI {
+ public:
+  static constexpr int kIndexPackedBuffer = 0;
+  static constexpr int kIndexPushConstant = 1;
+  static constexpr int kIndexWorkGroupId = 2;
+  static constexpr int kIndexWorkGroupCount = 3;
+  static constexpr int kIndexWorkGroupSize = 4;
+
+  // Returns the types of the LLVM function inputs for the ABI.
+  // This matches the signature of `iree_hal_executable_dispatch_v0_t` in
+  // `iree/hal/local/executable_library.h`.
+  static SmallVector<Type, 5> getInputTypes(MLIRContext *context) {
+    // func foo(%packed_buffer_args: !llvm.ptr<!llvm.ptr<i8>>,
+    //          %push_constant: !llvm.ptr<i32>,
+    //          workgroup_id[3]: !llvm.ptr<!llvm.array<i32, 3>>,
+    //          workgroup_count[3]: !llvm.ptr<!llvm.array<i32, 3>>,
+    //          workgroup_size[3]: !llvm.ptr<!llvm.array<i32, 3>>)
+    auto indexTy = IntegerType::get(context, 32);
+    return SmallVector<Type, 5>{
+        // %packed_buffer_args: !llvm.ptr<!llvm.ptr<i8>>
+        LLVM::LLVMPointerType::get(
+            LLVM::LLVMPointerType::get(IntegerType::get(context, 8))),
+        // %push_constant: !llvm.ptr<i32>
+        LLVM::LLVMPointerType::get(indexTy),
+        // %workgroup_id[3]: !llvm.ptr<!llvm.array<i32, 3>>
+        LLVM::LLVMPointerType::get(LLVM::LLVMArrayType::get(indexTy, 3)),
+        // %workgroup_count[3]: !llvm.ptr<!llvm.array<i32, 3>>
+        LLVM::LLVMPointerType::get(LLVM::LLVMArrayType::get(indexTy, 3)),
+        // %workgroup_size[3]: !llvm.ptr<!llvm.array<i32, 3>>
+        LLVM::LLVMPointerType::get(LLVM::LLVMArrayType::get(indexTy, 3))};
+  }
 
-// Return the index type: i32.
-static Type getIndexTy(MLIRContext *context) {
-  return IntegerType::get(context, 32);
-}
+  explicit HALDispatchABI(LLVM::LLVMFuncOp &funcOp,
+                          LLVMTypeConverter *typeConverter)
+      : funcOp(funcOp), typeConverter(typeConverter) {}
+
+  // Loads the workgroup_id[dim] value (XYZ) and casts it to |resultType|.
+  Value loadWorkgroupID(Location loc, int32_t dim, Type resultType,
+                        OpBuilder &builder) {
+    auto xyzArrayPtr = funcOp.getArgument(kIndexWorkGroupId);
```

Comment:
These three methods only differ in argument index, maybe we can have `Value loadIdArgument(int index, Location loc, int32_t dim, Type resultType, OpBuilder &builder)` and call it inside each of them

Reply:
The next PR #4791 changes that to not be the case - they are different :)